File: recipes/sequence.py
from __future__ import annotations

from dataclasses import dataclass

from utils.loader import GobiFile
from utils.recipes import GobiError, Action, Recipe


@dataclass(init=True)
class SequenceConfig:
    subactions: list[str]
    allow_fail: bool


class SequenceAction(Action):
    def __init__(self, name: str, config: SequenceConfig) -> None:
        self.name = "sequence." + name
        self.subname = name
        self.config = config

    def run(
        self,
        gobi_file: GobiFile,
        recipes: dict[str, Recipe],
        actions: list[Action],
        args: list[str],
    ) -> GobiError | None:
        actions_to_run: list[Action] = []
        for action in self.config.subactions:
            possible_actions = list(filter(lambda a: a.subname == action, actions))
            match possible_actions:
                case []:
                    possible_actions = list(filter(lambda a: a.name == action, actions))
                    match possible_actions:
                        case []:
                            return GobiError(self, 1, f"Unknown action: {action}")
                        case [possible_action]:
                            actions_to_run.append(possible_action)
                        case _:
                            return GobiError(self, 1, f"INTERNAL ERROR: Action {action} is ambiguous")
                case [possible_action]:
                    actions_to_run.append(possible_action)
                case _:
                    return GobiError(self, 1, f"Action {action} is ambiguous")

        errors = []
        for action in actions_to_run:
            res = action.run(gobi_file, recipes, actions, [])
            if self.config.allow_fail and res is not None:
                errors.append(res)
            elif res is not None:
                return GobiError(
                    self, res.code, f"Action {res.source.name} failed with msg:\n  {res.msg}"
                )
        if len(errors) == 1:
            return GobiError(
                self,
                errors[0].code,
                f"Action {res.source.name} failed with msg:\n  {errors[0].msg}",
            )
        elif len(errors) > 1:
            return GobiError(
                self,
                1,
                "Multiple actions failed with msgs:\n  {}".format(
                    "\n  ".join(map(lambda e: f"[{e.source.name}] ({e.code}): {e.msg}", errors))
                ),
            )


class SequenceRecipe(Recipe):
    def __init__(self):
        self.name = "sequence"

    def create_actions(self, gobi_file: GobiFile) -> GobiError | list[Action]:
        data = gobi_file.data.get("sequence", {})

        actions = []
        for action, action_data in data.items():
            sub_actions = action_data.get("subactions", [])
            allow_fail = action_data.get("allow_fail", False)
            config = SequenceConfig(sub_actions, allow_fail)
            actions.append(SequenceAction(action, config))
            
        return actions


def create() -> Recipe:
    return SequenceRecipe()

File: lib/recipes/src/sequence.rs
use serde::Deserialize;
use std::cell::OnceCell;

use gobi_lib::{
    file::{GobiFile, GobiFileEntryTrait, GobiFileTrait},
    recipes::*,
    GobiError, GobiResult,
};

#[derive(Deserialize)]
struct SequenceConfig {
    subactions: Vec<String>,
    #[serde(default)]
    #[serde(rename = "allow-fail")]
    allow_fail: bool,
    #[serde(default)]
    decorate: bool,
}

struct SequenceAction {
    name: String,
    subname: String,

    config: SequenceConfig,
    help: OnceCell<String>,
}

impl SequenceAction {
    fn new(name: &str, config: SequenceConfig) -> Self {
        SequenceAction {
            name: format!("sequence.{}", name),
            subname: name.to_string(),
            config,
            help: OnceCell::new(),
        }
    }
}

impl IAction for SequenceAction {
    fn get_name(&self) -> &str {
        &self.name
    }

    fn get_subname(&self) -> &str {
        &self.subname
    }

    fn get_help(&self) -> &str {
        self.help.get_or_init(|| {
            format!(
                "Runs {:?} in a sequence {}",
                self.config.subactions,
                if self.config.allow_fail {
                    "allowing failures"
                } else {
                    "exiting on failures"
                }
            )
        })
    }

    fn run(&self, actions: &Vec<Action>, args: Vec<String>) -> GobiResult<()> {
        let actions_to_run = self
            .config
            .subactions
            .iter()
            .map(|name| find_action(name, actions));

        let (err, actions_to_run) = actions_to_run.fold(
            (GobiError::default(), Vec::new()),
            |(err, mut actions), action| match action {
                Ok(action) => {
                    actions.push(action);
                    (err, actions)
                }
                Err(nerr) => (err.merge(nerr), actions),
            },
        );

        if err.code != 0 {
            return Err(err);
        }

        let err = actions_to_run
            .into_iter()
            .try_fold(GobiError::default(), |err, action| {
                if self.config.decorate {
                    let name = action.get_name();
                    println!("{0}\n# {1} #\n{0}", "#".repeat(name.len() + 4), name);
                }
                match action.run(actions, args.clone()) {
                    Ok(_) => Ok(err),
                    Err(nerr) if self.config.allow_fail => Ok(err.merge(nerr)),
                    Err(nerr) => Err(nerr),
                }
            });

        match err {
            Ok(err) if err.code == 0 => Ok(()),
            Ok(err) => Err(GobiError {
                code: err.code,
                msg: format!("Sequence finished with errors: "),
            }
            .merge(err)),
            Err(err) => Err(GobiError {
                code: err.code,
                msg: format!("Sequence failed: "),
            }
            .merge(err)),
        }
    }
}

struct SequenceRecipe {}

impl SequenceRecipe {
    fn new() -> SequenceRecipe {
        SequenceRecipe {}
    }
}

impl IRecipe for SequenceRecipe {
    fn get_name(&self) -> &str {
        "sequence"
    }

    fn get_help(&self) -> &str {
        "
Generates sequence actions.

This recipe uses the following configuration options:

[sequence.<action name>.subactions] (required) : list[str]
    list of actions to run in order

[sequence.<action name>.allow-fail] (optional) : bool (default: false)
    if true, continue running actions even if one fails

[sequence.<action name>.decorate] (optional) : bool (default: false)
    if true, prints the name of each subaction before running it
"
    }

    fn create_actions(
        &self,
        _recipe_manager: &SharedRecipeManager,
        gobi_file: &GobiFile,
    ) -> GobiResult<Vec<Action>> {
        let config = match gobi_file.get_data().get("sequence") {
            Some(config) if config.is_table() => config.get_table().unwrap(),
            _ => return Ok(vec![]),
        };

        let (err, actions) = config
            .into_iter()
            .map(
                |(name, config)| match config.deserialize::<SequenceConfig>() {
                    Ok(config) => {
                        Ok(ActionWrapper::new(SequenceAction::new(&name, config)) as Action)
                    }
                    _ => Err(GobiError {
                        code: 1,
                        msg: format!("Invalid sequence config for action '{}'", name),
                    }),
                },
            )
            .fold(
                (GobiError::default(), Vec::new()),
                |(err, mut actions), action| match action {
                    Ok(action) => {
                        actions.push(action);
                        (err, actions)
                    }
                    Err(nerr) => (err.merge(nerr), actions),
                },
            );

        if err.code != 0 {
            return Err(err);
        }

        Ok(actions)
    }
}

pub fn register_recipes(recipe_manager: &mut RecipeManager) {
    recipe_manager.register_recipe(RecipeWrapper::new(SequenceRecipe::new()));
}
